pkg/querier/engine/streaming/operator/selector.go

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operator
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+type Selector struct {
+	Queryable storage.Queryable
+	Start     int64 // Milliseconds since Unix epoch
+	End       int64 // Milliseconds since Unix epoch
+	Timestamp *int64
+	Interval  int64 // In milliseconds
+	Matchers  []*labels.Matcher
+
+	// Set for instant vector selectors, otherwise 0.
+	LookbackDelta time.Duration
+
+	// Set for range vector selectors, otherwise 0.
+	Range time.Duration
+
+	querier                   storage.Querier
+	currentSeriesBatch        *seriesBatch
+	seriesIndexInCurrentBatch int
+}
+
+// There's not too much science behind this number: this is based on the batch size used for chunks streaming.
+const seriesBatchSize = 256
+
+var seriesBatchPool = sync.Pool{New: func() any {
+	return &seriesBatch{
+		series: make([]storage.Series, 0, seriesBatchSize),
+		next:   nil,
+	}
+}}
+
+func (s *Selector) SeriesMetadata(ctx context.Context) ([]SeriesMetadata, error) {
+	if s.currentSeriesBatch != nil {
+		return nil, errors.New("should not call Selector.SeriesMetadata() multiple times")
+	}
+
+	if s.LookbackDelta != 0 && s.Range != 0 {
+		return nil, errors.New("invalid Selector configuration: both LookbackDelta and Range are non-zero")
+	}
+
+	startTimestamp := s.Start
+	endTimestamp := s.End
+
+	if s.Timestamp != nil {
+		startTimestamp = *s.Timestamp
+		endTimestamp = *s.Timestamp
+	}
+
+	rangeMilliseconds := s.Range.Milliseconds()
+	start := startTimestamp - s.LookbackDelta.Milliseconds() - rangeMilliseconds
+
+	hints := &storage.SelectHints{
+		Start: start,
+		End:   endTimestamp,
+		Step:  s.Interval,
+		Range: rangeMilliseconds,
+		// TODO: do we need to include other hints like Func, By, Grouping?
```

Comment:
Some comments about Mimir:

- We don't use Grouping, By, Range. Unless we want to use them in the streaming engine, we don't need to value them.
- We use `Func` to determine whether it's a `/series` request, but it doesn't go through the PromQL engine so doesn't affect this
- What's about `ShardCount`, `ShardIndex` and `DisableTrimming` which we use in Mimir? `ShardCount` and `ShardIndex` shouldn't be used at this level, because sharding label matcher is passed down to ingesters and store-gateway, then they filter it out and return series only matching to the requested shard. Same goes with `DisableTrimming` which is just set to true in the ingesters, when querying TSDB.

So, if I understand correctly your TODO, I think we can remove it for now.

Reply:
Thanks for the clarification, I've replaced the TODO with your explanation in 95699d6.